### PR TITLE
Shareable Add-ons validations

### PR DESCRIPTION
**Pull request by mnoble**
Adds some Manifest and ProvisionResponse checks specific to Shareable Add-ons. It might be easier to go commit by commit, just a heads up.

cc @heroku/add-ons

### CODE REVIEW

File: lib/heroku/kensa/check.rb [hunk 1]
```diff
@@ -173,6 +173,14 @@ def call!
           response.is_a?(Hash) && response.has_key?("id")
         end
 
+        check "id does not contain heroku_id" do
+          if response["id"] && response["id"].respond_to?(:include?) && response["id"].include?(data["heroku_id"])
```
Comment (bjeanes) [inline lib/heroku/kensa/check.rb]: `reponse["id"]` is mandatory (if this isn't tested, we should add that too), so it should always be here and non-blank.

It should also always be a string, too. However, for legacy reasons there may be some who sent us ints, so we coerce it to string in the back end. I'd just coerce `to_s` then check if it includes the ID.

Lastly, we don't want to check if it just includes `heroku_id`. Some strip it or regex out the number from it. We should check if the numeric ID portion is present, IMO

Comment (mnoble) [inline lib/heroku/kensa/check.rb]: re: `response["id"]` & string vs int – I was mostly going off of the other tests. There's some that use ints and some that return a `{ id: nil }` which the check above this one would not catch.
re: heroku_id – I took "heroku id" in the TP card to mean the hid.

I'll update everything to match what what you've mentioned. Thanks!

Comment (bjeanes) [inline lib/heroku/kensa/check.rb]: > I took "heroku id" in the TP card to mean the hid.

Sorry, could have been clearer. I have a dataclip somewhere and I saw some IDs were: `heroku_app125123` which is also problematic, so testing for the numeric part is hopefully sufficient.

>  some that return a `{ id: nil }`

Yikes. Our actual provisioning code definitely expects an `id` 100% of the time for provision requests. We should add a test for that if there isn't already.

**Comment by bjeanes** [comment]
Thanks for those changes. I haven't grokked the rest of the PR yet. I saw something like you were booting up a little server and I closed the tab for later :|

**Comment by bjeanes** [comment]
A support ticket just pointed out that Kensa doesn't send the resource UUID either. We'll need to update these specs accordingly.

**Comment by mnoble** [comment]
@bjeanes It was missing `name` as well, so I added that.

**Comment by bjeanes** [comment]
Do we send the `name`... We do for the HPG version but doing so was a mistake since it is mutable. I am reasonably sure we don't for the v1 integration.

**Comment by mnoble** [comment]
Ahh yea, `v2` does, but `v1` does not.

File: lib/heroku/kensa/check.rb [hunk 2]
```diff
@@ -290,6 +298,73 @@ def call!
...
+            if json1["id"] == json2["id"]
+              error "multiple provisions cannot return the same id"
+            else
+              true
+            end
+          end
+        else
+          check "disallows duplicate provisions" do
+            if code2 == 200
+              error("should not allow duplicate provisions")
```
Comment (bjeanes) [inline lib/heroku/kensa/check.rb]: I don't think we should have this test. This kind of logic is the cause behind some provision errors due to race conditions that occur when a customer uninstalls something and quickly re-installs it (i.e. before the deprovision request has occurred). I think it should probably always succeed with multiple provisions of the same ID. What do you think?

File: lib/heroku/kensa/check.rb [hunk 3]
```diff
@@ -290,6 +298,73 @@ def call!
...
+            reader.close
+            server = TCPServer.open(7779)
+            client = server.accept
+            writer.write(client.readpartial(READLEN))
+            client.write("HTTP/1.1 200 OK\r\nContent-Length: 0\r\n\r\n")
+            client.close
+            writer.close
+          end
+          sleep(1)
+        end
```
Comment (bjeanes) [inline lib/heroku/kensa/check.rb]: @mnoble I don't understand what this block is doing. Can you help me understand it?

Comment (mnoble) [inline lib/heroku/kensa/check.rb]: uhhhh, not really, haha. I brought it over from `ProvisionCheck`. I'm actually not sure if it's even needed or does anything in `ProvisionCheck`. I'm going to remove it from here.

File: lib/heroku/kensa/check.rb (hunk 2 above)
Comment (mnoble) [inline lib/heroku/kensa/check.rb]: As in, we should encourage providers **without** `many_per_app` to make this endpoint idempotent?

Comment (bjeanes) [inline lib/heroku/kensa/check.rb]: Not idempotent (that would imply returning the same provision/account details) but it shouldn't be an error condition. They should always provision as many add-ons as requests we make, under the premise that the first one may be pending delete. 

There's a counter argument here though: when receiving a duplicate request, they could check the App Info API to see if the other one is still installed and reject on that premise. However, this test might encourage people to too aggressively reject valid installations.

Comment (mnoble) [inline lib/heroku/kensa/check.rb]: Ahh, okay, makes sense.

**Comment by bjeanes** [comment]
hoorah. Can you cut a release?

**Comment by mattreduce** [comment]
@bjeanes I can tag and release now. I'm leaning towards tagging this `2.3.0`. Would you say that's appropriate, or has this changed enough/in a backwards-incompatible way that warrants a version `3.0.0`?

**Comment by bjeanes** [comment]
@mattreduce it's not a library so much as a test suite the semantic versioning rules are hard to map cleanly. Either seems fine to me.

**Comment by mattreduce** [comment]
@bjeanes :psyduck: 2.3.0 it is

**Comment by mattreduce** [comment]
:ship: kensa 2.3.0 is out https://rubygems.org/gems/kensa/versions/2.3.0

**Comment by mnoble** [comment]
My bad. I merged it and forgot to push the gem. Thanks a ton @mattreduce !

**Pull request merged**